### PR TITLE
Connecticut Credit based on AGI

### PR DESCRIPTION
**Pull request by ThapeloMasasa**
Added the Parameter file for Head of household brackets and Joint. if approved then the other categories to be added

### CODE REVIEW

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml:1]: we need a total of 5 parameter files - single, joint, separate, head_of_household and widow

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml [hunk 1]
```diff
@@ -0,0 +1,128 @@
+description: These are the brackets for married individuals filing jointly.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml:1]:
```suggestion
description: Connecticut provides the following credit amount under the AGI credit, based state AGI for joint filers.
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml:1]: similar in all yaml

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+metadata:
+  type: marginal_rate 
+  threshold_unit: currency-USD
+  rate: /1
+  label: Connecticut married Tax Brackets for adjusted gross income
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml:6]:
```suggestion
  label: Connecticut AGI credit joint filers amount
```

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  reference:
+    - title: Connecticut General Statutes, Chapter 229, Sec. 12-703, (1), (I)
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml:8]: need to add 2022 tax form reference as well

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  
+
+
+    
+
+    
+
+
+    
+
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml:128]: delete extra lines

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    rate:
+      2016-01-01: 0.03
+  - threshold:
+      2016-01-01: 99_500
+    rate:
+      2016-01-01: 0.02
+  - threshold:
+      2016-01-01: 100_000
+    rate:
+      2016-01-01: 0.01
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml:118]: last bracket missing - should be a rate of 0 as final

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  type: marginal_rate 
+  threshold_unit: currency-USD
+  rate: /1
+  label: Connecticut married Tax Brackets for adjusted gross income
+  reference:
+    - title: Connecticut General Statutes, Chapter 229, Sec. 12-703, (1), (I)
+      href: https://www.cga.ct.gov/current/pub/chap_229.htm#sec_12-703
+brackets:
+  - threshold: 
+      2016-01-01: 24_000
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml:12]: is the rate above 0 or 1?

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml [hunk 7]
```diff
@@ -0,0 +1,134 @@
+description: Connecticut provides the following credit amount under the AGI credit, based state AGI for head of household filers.
+metadata:
+  type: marginal_rate 
+  threshold_unit: currency-USD
+  rate: /1
+  label: AGI credit head of household filers amount
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml:6]:
```suggestion
  label: Connecticut AGI credit head of household filers amount
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml:6]: all yaml

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+  reference:
+    - title: Connecticut General Statutes, Chapter 229, Sec. 12-703, (1), (I)
+      href: https://www.cga.ct.gov/current/pub/chap_229.htm#sec_12-703
+    - title: 2022 Form CT-1040 Connecticut Resident Income Tax Return Instructions
+      href: https://portal.ct.gov/-/media/DRS/Forms/2022/Income/2022-CT-1040-Instructions_1222.pdf
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml:11]:
```suggestion
      href: https://portal.ct.gov/-/media/DRS/Forms/2022/Income/2022-CT-1040-Instructions_1222.pdf#page=
```
insert page number

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml:11]: all yaml

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml [hunk 9]
```diff
@@ -0,0 +1,124 @@
+description: Connecticut provides the following credit amount under the AGI credit, based state AGI for joint filers.
+metadata:
+  type: single_value 
+  threshold_unit: currency-USD
+  rate: /1
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml:5]:
```suggestion
  rate_unit: /1
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml:5]: all yaml

File: policyengine_us/variables/gov/states/ct/tax/income/credits/ct_credit_based_on_agi.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
+from policyengine_us.model_api import *
+
+
+class ct_credit_based_on_agi(Variable):
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ct/tax/income/credits/ct_credit_based_on_agi.py:4]:
```suggestion
class ct_personal_credits(Variable):
```

File: policyengine_us/variables/gov/states/ct/tax/income/credits/ct_credit_based_on_agi.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    value_type = float
+    entity = TaxUnit
+    label = "Connecticut Credit Based on AGI"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ct/tax/income/credits/ct_credit_based_on_agi.py:7]:
```suggestion
    label = "Connecticut Personal Credits"
```

File: policyengine_us/variables/gov/states/ct/tax/income/credits/ct_credit_based_on_agi.py (hunk 10 above)
Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/ct/tax/income/credits/ct_credit_based_on_agi.py:4]: need to rename for alignment

File: changelog_entry.yaml
Comment (PavelMakarchuk) [inline changelog_entry.yaml:1]: run `make format`

Comment (PavelMakarchuk) [inline changelog_entry.yaml:1]: run `git pull upstream master`

**Review by PavelMakarchuk — APPROVED** [review]

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+metadata:
+  type: single_amount 
+  threshold_unit: currency-USD
+  amount_unit: /1
+  label: Connecticut AGI credit head of household filers amount
+  reference:
+    - title: Connecticut General Statutes, Chapter 229, Sec. 12-703, (1), (I)
+      href: https://www.cga.ct.gov/current/pub/chap_229.htm#sec_12-703
+    - title: 2022 Form CT-1040 Connecticut Resident Income Tax Return Instructions
+      href: https://portal.ct.gov/-/media/DRS/Forms/2022/Income/2022-CT-1040-Instructions_1222.pdf#page=23
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml:11]: link doesn't work

File: policyengine_us/variables/gov/states/ct/tax/income/credits/ct_personal_credits.py [hunk 13]
```diff
@@ -0,0 +1,32 @@
+from policyengine_us.model_api import *
+
+
+class ct_personal_credits(Variable):
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/ct/tax/income/credits/ct_personal_credits.py:4]: something named this way would be a dollar-value credit. this needs to be multiplied by something to get that

File: changelog_entry.yaml [hunk 14]
```diff
@@ -0,0 +1,4 @@
+- bump: minor
+  changes:
+    added:
+    - Connecticut credit based on agi
```
Comment (MaxGhenis) [inline changelog_entry.yaml:4]:
```suggestion
    - Connecticut credit based on AGI.
```

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml [hunk 15]
```diff
@@ -0,0 +1,134 @@
+description: Connecticut provides the following credit amount under the AGI credit, based state AGI for head of household filers.
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml:1]: I'm guessing here, but in general the description should express what the values are and not abbreviate
```suggestion
description: Connecticut provides head of household filers a credit for this fraction of taxes, depending on their adjusted gross income.
```

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+metadata:
+  type: single_amount 
+  threshold_unit: currency-USD
+  amount_unit: /1
+  label: Connecticut AGI credit head of household filers amount
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml:6]:
```suggestion
  label: Connecticut AGI credit for head of household filers
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/ct/tax/income/credits/ct_personal_credits.py (hunk 13 above)
Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/ct/tax/income/credits/ct_personal_credits.py:4]: they call it the personal credit as it is used in computing in the[ income tax](https://github.com/PolicyEngine/policyengine-us/pull/2586) PR we would not multiply it in this variable but in the ct_income_tax_after_personal_credits variable

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+    amount:
+      2016-01-01: 0.02
+  - threshold:
+      2016-01-01: 100_000
+    amount:
+      2016-01-01: 0.01
+  - threshold:
+      2016-01-01: 100_500
+    amount:
+      2016-01-01: 0.00
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/joint.yaml:124]:
```suggestion
      2016-01-01: 0
```

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/widow.yaml [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+    amount:
+      2016-01-01: 0.02
+  - threshold:
+      2016-01-01: 100_000
+    amount:
+      2016-01-01: 0.01
+  - threshold:
+      2016-01-01: 100_500
+    amount:
+      2016-01-01: 0.00
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/widow.yaml:124]:
```suggestion
      2016-01-01: 0
```

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/single.yaml [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+    amount:
+      2016-01-01: 0.02
+  - threshold:
+      2016-01-01: 64_000
+    amount:
+      2016-01-01: 0.01
+  - threshold:
+      2016-01-01: 64_500
+    amount:
+      2016-01-01: 0.00
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/single.yaml:124]:
```suggestion
      2016-01-01: 0
```

File: policyengine_us/variables/gov/states/ct/tax/income/credits/ct_personal_credits.py (hunk 13 above)
Comment (MaxGhenis) [reply policyengine_us/variables/gov/states/ct/tax/income/credits/ct_personal_credits.py:4]: OK then let's rename to `ct_personal_credit_rate` and add `unit = "\1"`

File: policyengine_us/variables/gov/states/ct/tax/income/credits/ct_personal_credits.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ct/tax/income/credits/ct_personal_credits.py:1]: rename file to ct_personal_credit_rate.py

File: policyengine_us/tests/policy/baseline/gov/states/ct/tax/income/credits/ct_personal_credits.yaml
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/ct/tax/income/credits/ct_personal_credits.yaml:1]: ct_personal_credit_rate.yaml rename to

File: policyengine_us/tests/policy/baseline/gov/states/ct/tax/income/credits/ct_personal_credits.yaml [hunk 20]
```diff
@@ -0,0 +1,45 @@
+- name: Single household with $70,900 agi
+  period: 2016
+  input:
+    state_code: CT
+    ct_agi: 70_900
+    filing_status: SINGLE
+  output:
+    ct_personal_credits: 0
```
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/ct/tax/income/credits/ct_personal_credits.yaml:8]:
```suggestion
    ct_personal_credit_rate: 0 
```

Comment (PavelMakarchuk) [reply policyengine_us/tests/policy/baseline/gov/states/ct/tax/income/credits/ct_personal_credits.yaml:8]: and all others

File: policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+      2016-01-01: 0.02
+  - threshold:
+      2016-01-01: 78_000
+    amount:
+      2016-01-01: 0.01
+  - threshold:
+      2016-01-01: 78_500
+    amount:
+      2016-01-01: 0
+
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/ct/tax/income/credits/agi/head_of_household.yaml:125]: remove empty lines

File: policyengine_us/variables/gov/states/ct/tax/income/credits/ct_personal_credit_rate.py [hunk 22]
```diff
@@ -0,0 +1,33 @@
+from policyengine_us.model_api import *
+
+
+class ct_personal_credit_rate(Variable):
+    value_type = float
+    entity = TaxUnit
+    unit = "\1"
+    label = "Connecticut Personal Credits"
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/ct/tax/income/credits/ct_personal_credit_rate.py:8]:
```suggestion
    label = "Connecticut personal credit rate"
```

**Review by MaxGhenis — APPROVED** [review]
good to merge after these

**Pull request merged**